src/ray/core_worker/transport/direct_task_transport.cc

```diff
@@ -606,16 +607,9 @@ void CoreWorkerDirectTaskSubmitter::PushNormalTask(
           if (reply.worker_exiting()) {
             RAY_LOG(DEBUG) << "Worker " << addr.worker_id
                            << " replied that it is exiting.";
-            // The worker is draining and will shutdown after it is done. Don't return
-            // it to the Raylet since that will kill it early.
-            worker_to_lease_entry_.erase(addr);
-            auto &scheduling_key_entry = scheduling_key_entries_[scheduling_key];
-            scheduling_key_entry.active_workers.erase(addr);
-            if (scheduling_key_entry.CanDelete()) {
-              // We can safely remove the entry keyed by scheduling_key from the
-              // scheduling_key_entries_ hashmap.
-              scheduling_key_entries_.erase(scheduling_key);
-            }
+            // Since this worker is draining, make sure the raylet doesn't
```

Comment:
It seems we do :(. This fixes a bug where worker processes weren't returned they were killed (which could be quite late since they don't die until their ref counts go away). It seems we most likely didn't notice before because we did unblock the worker, which freed the CPUs.

Reply:
I'm not sure I understand why this is needed. If the worker is exiting anyway, why would it matter if we return it or not? I also don't understand what this has to do with the rest of the PR. Does it make sense to separate this part out and test it separately?